File: leecode/leetcode-cn/动态规划/887.鸡蛋掉落.py
# https://www.cnblogs.com/xym4869/p/12581145.html
// https://leetcode-cn.com/problems/super-egg-drop/solution/887-by-ikaruga/
# labuladong

# dp: 设 f(k,n) 为测试n层楼，当有k个鸡蛋时，最坏情况下，最少需要测试的次数。
# (1) 碎了，小明还剩下k-1个鸡蛋，i-1 层待测试楼层，于是问题变为 f(k-1, i-1)
# (2) 没碎，小明还剩下k个鸡蛋，n-i 层待测试楼层，于是问题变为 f(k, n-i)
# 公式中的"+1" : 因为第i层测试了一次

# @ref: labuladong
def superEggDrop(K: int, N: int) -> int:
    memo = dict() # 备忘录去重
    def dp(K, N):
        res = float('INF') # 如果在第一层楼碎了，那么res可以取0
        if  K==1: # K从1开始，因为K==1时，结果是确定的
            return N
        if N==0: # N从0开始，因为N==0时，结果是确定的；N==1时，结果不确定
            return 0
        # 避免重复计算
        if (K,N) in memo:
            return memo[(K,N)]
        # 穷举所有可能的选择
        for i in range(1,N+1):
            res = min(
                max(dp(K-1,i-1),dp(K,N-i))+1,
                res
            )
        # 记入备忘录
        memo[(K,N)]=res
        return res
    return dp(K,N)

if __name__ == "__main__":
    print(superEggDrop(2,100))

File: leecode/leetcode-cn/动态规划/887.鸡蛋掉落.cc
// https://www.cnblogs.com/xym4869/p/12581145.html
// labuladong

/**
 * @brief 双蛋问题
 *        2个鸡蛋，100层楼，找出扔蛋恰好不碎的楼层，求最少测试次数
 * 
 * @ans 这是求最坏情况下的最少次数
 *      贪心算法，假设最优解为x，每次扔蛋都保证最优。
 *      第1次扔，剩下次数为x-1，最坏情况下，该蛋碎了，那么剩下的区间中只能线性探测，也就是该区间长度不得超过x-1
 *      第2次扔，需要探测的区间长度不得超过x-2，最多还需测试x-2次
 *      ...
 *      第x次扔，剩余区间为0
 *      于是，有 1+2+...+x >= 100，求最小的x
 *      => x = 14
 * 
 *     也就是，第一次在14层扔，第二次在14+13=27层扔，...
 */


// 作者：ikaruga
// 链接：https://leetcode-cn.com/problems/super-egg-drop/solution/887-by-ikaruga/
class Solution {
public:
    int calcF(int K, int T)
    {
        if (T == 1 || K == 1) return T + 1;
        return calcF(K - 1, T - 1) + calcF(K, T - 1);
    }

    int superEggDrop(int K, int N)
    {
        int T = 1;
        while (calcF(K, T) < N + 1) T++;
        return T;
    }
};
